content/en/blog/_posts/2025-10-25-new-cgroup-v1-to-v2-conversion-formula.md

```diff
@@ -0,0 +1,179 @@
+introduced a conversion formula to map cgroup v1 CPU shares to cgroup v2 CPU 
+weight. The conversion formula was defined as:
+
+```
+cpu.weight = (1 + ((cpu.shares - 2) * 9999) / 262142)
+```
+
+<!-- convert from [2-262144] to [1-10000] -->
+
+This formula linearly maps between \([2^1 - 2^18]\) to \([10^0 - 10^4]\).
+![2025-10-25-new-cgroup-v1-to-v2-conversion-formula-linear-conversion.png](2025-10-25-new-cgroup-v1-to-v2-conversion-formula-linear-conversion.png)
+
+While this approach is simple, the linear mapping imposes a few significant 
+problems and impacts both performance and configuration granularity.
+
+## Problems with Previous Conversion Formula
+
+The current conversion formula creates two major issues:
+
+### 1. Reduced Priority Against Non-Kubernetes Workloads
+
+In cgroup v1, the default CPU shares is `1024`, meaning a container 
+requesting 1 CPU has equal priority with system processes that live outside 
+of Kubernetes' scope.
+However, in cgroup v2, the default CPU weight is `100`, but the current 
+formula converts 1 CPU (1024m) to only `~39` weight - less than 40% of the 
+default.
+
+**Example:**
+- Container requesting 1 CPU (1024m)
+- cgroup v1: `cpu.shares = 1024` (equal to default)
+- cgroup v2 (current): `cpu.weight = 39` (much lower than default 100)
+
+This means that after moving to cgroup v2, Kubernetes workloads would 
+de-factor reduce their CPU priority against non-Kubernetes processes. The 
+problem can be severe for setups with many system daemons that run 
+outside of Kubernetes' scope and expect Kubernetes workloads to have 
+priority, especially in situations of resource starvation.
+
+### 2. Unmanageable Granularity
+
+The current formula produces very low values for small CPU requests, 
+limiting the ability to create sub-cgroups within containers for 
+fine-grained resource distribution (which will possibly be much easier moving
+forward, see [KEP #5474](https://github.com/kubernetes/enhancements/issues/5474) for more info).
+
+**Example:**
+- Container requesting 100m CPU
+- cgroup v1: `cpu.shares = 102`
+- cgroup v2 (current): `cpu.weight = 4` (too low for sub-cgroup 
+  configuration)
+
+With cgroup v1, requesting 1 CPU which led to 102 CPU shares was manageable 
+in the sense that sub-cgroups could have been created inside the main 
+container, assigning fine-grained CPU priorities for different groups of 
+processes. With cgroup v2 however, having 4 shares is very hard to 
+distribute between sub-cgroups since it's not granular enough.
+
+With plans to allow [writable cgroups for unprivileged containers](https://github.com/kubernetes/enhancements/issues/5474),
+this becomes even 
+more relevant.
+
+## New Conversion Formula
+
+### Description
+The new formula is more complicated, but does a much better job mapping 
+between cgroup v1 CPU shares and cgroup v2 CPU weight:
+
+```
+cpu.weight = ⌈10^(L²/612 + 125L/612 - 7/34)⌉
+```
```

Comment:
Math mode can be used here?

https://www.docsy.dev/docs/adding-content/diagrams-and-formulae/